Sources/EdgeNetworkHandlers/ResponseCallbackHandler.swift

```diff
@@ -39,32 +51,93 @@ class ResponseCallbackHandler {
         responseHandlers[requestEventId] = unwrappedResponseHandler
     }
 
-    /// Unregisters a `EdgeResponseHandler` for the specified `requestEventId`. After this operation,
-    /// the associated response handler is not invoked anymore.
+    /// Registers a completion handler for the specified `requestEventId`. This handler is invoked when the entire Edge response content has been
+    /// handled entirely by the Edge extension, containing a list of `EdgeEventHandle` and a list of `EdgeEventError`. These lists can be empty or
+    /// can contain one or multiple items based on the request and the server side response.
+    ///
+    /// - Parameters:
+    ///   - requestEventId: unique event identifier for which the response callback is registered; should not be empty
+    ///   - completionHandler: the completion handler that needs to be registered, should not be nil
+    func registerCompletionHandler(requestEventId: String, completionHandler: (([EdgeEventHandle], [EdgeEventError]) -> Void)?) {
+        guard let unwrappedCompletion = completionHandler else { return }
+        guard !requestEventId.isEmpty else {
+            Log.warning(label: TAG, "Failed to register completion handler because of empty request event id.")
+            return
+        }
+
+        completionHandlers[requestEventId] = unwrappedCompletion
+    }
+
+    /// Calls onComplete and unregisters a `EdgeResponseHandler` for the specified `requestEventId`. After this operation,
+    /// the associated response handler or completion handler is removed and no longer called.
     /// - Parameter requestEventId: unique event identifier for experience events; should not be empty
-    func unregisterResponseHandler(requestEventId: String) {
+    func unregisterCallbacks(requestEventId: String) {
         guard !requestEventId.isEmpty else { return }
-
-        if responseHandlers[requestEventId] != nil {
+        if let responseHandler = responseHandlers[requestEventId] {
+            responseHandler.onComplete()
             responseHandlers[requestEventId] = nil
-            Log.trace(label: TAG, "Removing callback for Edge response with request unique id \(requestEventId).")
         }
+
+        if let completionHandler = completionHandlers[requestEventId] {
+            completionHandler(edgeEventHandles[requestEventId] ?? [], edgeEventErrors[requestEventId] ?? [])
+            completionHandlers[requestEventId] = nil
+        }
+
+        edgeEventHandles[requestEventId] = nil
+        edgeEventErrors[requestEventId] = nil
```

Comment:
When clearing values from the ThreadSafeDictionary objects, can you use `removeValue` instead of setting it to `nil`? My thinking is setting the value to `nil` will keep the element in the dictionary and not free up memory.

Reply:
ThreadSafeDictionary is using a Dictionary under the hood, so setting the value to nil is removing the key-value pair from the dictionary, as stated in the swift doc https://developer.apple.com/documentation/swift/dictionary. You can find more discussions on this thread https://forums.swift.org/t/dictionary-with-optional-values/2675/15.
I updated the implementation though so that it is independent on the internal implementation of ThreadSafeDictionary.